### PR TITLE
Typo

### PR DESCRIPTION
**Pull request by PallHaraldsson**

### CODE REVIEW

**Comment by PallHaraldsson** [comment]
Hi, at least PNSR was a typo, but note reading from the graph, for Julia seemed less than 10 sec. I suppose should have been five. And if it was off by ten for some reason, ten also assume 2.6. Do not merge if I'm wrong.

**Comment by PallHaraldsson** [comment]
FYI, there is from 2019:

An Improved WNNM Algorithm for Image Denoising
https://iopscience.iop.org/article/10.1088/1742-6596/1237/2/022037
>The traditional weighted nuclear norm minimization (WNNM) has excellent performance for the removal of non-sparse noise such as Gaussian noise, but attains bad performance for the removing of salt&pepper noise [..]
This paper proposes an improved WNNM image denoising algorithm. [..]
The experimental data demonstrate that the improved WNNM algorithm has better denoising effect than WNNM algorithm.

I was going to point to your package at: https://discourse.julialang.org/t/is-there-julia-library-for-ai-image-denoising/75688

but first read up on it, see if it's "AI denoising". I see it was state-of-the-art at the time, in 2014, and while imported upon in 2019, it's not clear to me it's still best, nor "AI", even though that or neural networks aren't a goal in of itself. I'm now kind of curious how fast algorithm is (since it's supposed to work for video), or the alternative SOTA I point to (i.e. there under my nickname Palli) at the thread.

Also if I should know more about this algorithm I hadn't heard of, even if for some other applications, not just denoising:

https://arxiv.org/abs/1405.6012

https://arxiv.org/abs/1608.04517
>In recent years, various applications based on
NNM have been developed, such as image/video denoising
[1, 4], background extraction [5, 6] and face shadow removal
[13, 14].

**Comment by PallHaraldsson** [comment]
FYI: If you're still interested in denoising, I believe this is the current SOTA for it (also very intriguing to me for other reasons):

Simple Baselines for Image Restoration
https://arxiv.org/pdf/2204.04676.pdf
>In this paper, we propose a simple baseline that exceeds the SOTA methods and is computationally efficient. To further simplify the baseline, we reveal that the nonlinear activation functions, e.g. Sigmoid, ReLU, GELU, Softmax, etc. are not necessary: they could be replaced by multiplication or removed. Thus, we derive a Nonlinear Activation Free Network, namely NAFNet, from the baseline. SOTA results are achieved on various challenging benchmarks, e.g. 33.69 dB PSNR on GoPro (for image deblurring), exceeding the previous SOTA 0.38 dB with only 8.4% of its computational costs; 40.30 dB PSNR on SIDD (for image denoising), exceeding the previous SOTA 0.28
dB with less than half of its computational costs. [..]
>
>Image restoration tasks aim to restore a degraded image (e.g. noisy, blur) to a clean one. Recently, deep learning based methods[4,35,37,34,5,6,30,7,24] achieve SOTA results on these tasks, and most of the methods could be viewed as variants of a classical solution, UNet[28]. [..]
>
>3.5 Attention
>Considering the recent popularity of the transformer in computer vision, its
attention mechanism is an unavoidable topic in the design of the internal structure of the block. [..]
The vanilla self-attention mechanism[32], which is adopted by [11,3], [..] suffers from the quadratic computational complexity with the size of the feature map. Some image restoration tasks process data at high resolution which makes the vanilla self-attention not practical. Alternatively, [21,20,34] apply self-attention only in a fix-sized local
window to alleviate the issue of increased computation. While it lacks global information. We do not take the window-based attention

It's not obvious to me that you need global information... noise is local, right, and it should be enough to split the image into many "fix-sized local" (maybe overlapping) windows and denoise them separately.

**Comment by johnnychen94** [comment]
> note reading from the graph, for Julia seemed less than 10 sec. I suppose should have been five. And if it was off by ten for some reason, ten also assume 2.6. Do not merge if I'm wrong.

Ha! I believe the 50s and 26s are the time without multi-threading enabled. I've fixed the typo directly in the master branch.

> Simple Baselines for Image Restoration

I also noticed this paper and had planned to reproduce it when I get some time. Having a Julia AI-based network as a baseline is a definitely something I want for a long time.

My current focus is still on patch-based analysis, on which I agree with you, that I also believe local patches already give enough information to denoise well.

---

FWIW, the patch-based WNNM denoiser implemented here actually solves a small WNNM problem **per pixel**, so to denoise a 512x512 image, you'll have to solve 256^2 (say, stride=2) WNNM problems.
This fact makes it way sloooow to use it in practice.
